Fix Jetpack standalone only sites visible on wpcom site selector

manzoorwanijk opened the pull request:
Sites with Jetpack standalone sites do not work well in WPCOM because WPCOM expects REST API to be available for the site which is possible only via the full Jetpack plugin. Thus, many UI sections on WPCOM for sites with standalone Jetpack plugins break e.g. Users and Stats section for a site with Standalone Backup, Search or Boost.

#### Proposed Changes

* This PR filters out the sites with Jetpack standalone plugins from WPCOM site selector
* It retains those sites in site selector for Jetpack Cloud
* It extracts `isJetpackSitePred` predicate from `isJetpackSite` selector to make it re-usable when we already have a site object.

#### Testing Instructions

- Create a site (JN site) with only a standalone Jetpack plugin, e.g., Backup, Search, Boost, etc.
- Boot up this PR and run `yarn start` and also `start-jetpack-cloud-p` in another terminal window/tab
- Connect Jetpack on the site with an appropriate plan
- If you see errors during connection, then replace `https://wordpress` with `http://calypso.localhost:3000` in the connection URL.
- Goto [http://calypso.localhost:3000/sites](http://calypso.localhost:3000/sites)
- Confirm that the JN site with only standalone Jetpack plugin IS NOT available in the selector
- Goto [http://jetpack.cloud.localhost:3001/landing](http://jetpack.cloud.localhost:3001/landing)
- Confirm that the site with only standalone Jetpack plugin IS available in the selector
- Select the site
- Confirm that everything works fine

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes 1164141197617539-as-1202835600809449/f

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1202835600809449

manzoorwanijk committed:
Extract isJetpackSitePred predicate from isJetpackSite

manzoorwanijk committed:
Update filters for site selector to filter out standalone sites in WPCOM

manzoorwanijk committed:
Revert the change made in #66828 which is not needed now

a8ck3n commented:
The changes to the site selector are working for me. I tested in Calypso Green (site is visible in selector) and Calypso Blue (not in the site selector). I wasn't able to get to the Atomic checkout tests though, as I need to go offline.

Of note, I can still view the site in Calypso Blue if I manually update the address field. ie:

![BlueSnap01](https://user-images.githubusercontent.com/40267301/186125436-6fa9135b-338f-4713-95f7-580eaf12b02b.png)

In this way, we can still run into errors on pages like Users, though I guess this is not a very likely scenario.

![BlueSnap02](https://user-images.githubusercontent.com/40267301/186125730-fb6b8a08-22d3-4f32-9e9a-b0edcc11a236.png)

Not something to be addressed here but possibly worth a lower-priority follow-up issue to make the effected pages aware of Jetpack-connected sites that do not have a the full plugin installed.

manzoorwanijk commented:
> Not something to be addressed here but possibly worth a lower-priority follow-up issue to make the effected pages aware of Jetpack-connected sites that do not have a the full plugin installed.

Thank you @a8ck3n for testing this. Yes, I agree that such guards should be added in some future PR.

manzoorwanijk committed:
Enhance isJetpackSitePred to take care of atomic sites

manzoorwanijk committed:
Create getJetpackActivePlugins helper function

manzoorwanijk committed:
Fix the issue of simple sites being filtered out

mattgawarecki commented:
@manzoorwanijk My experience mirrors that of @a8ck3n, but when I used an Atomic site to check for regressions to #66828, the button to upgrade Jetpack Search wasn't visible in the UI. I'm not yet sure why this is, but here are some screenshots to demonstrate what I'm seeing:

##### This PR @ `/settings/performance/<mysite>`

<img width="745" alt="image" src="https://user-images.githubusercontent.com/670067/186261147-b6857713-7cf8-46b9-bc9f-eee3326f157a.png">

##### WordPress.com staging @ `/settings/performance/<mysite>`

<img width="748" alt="image" src="https://user-images.githubusercontent.com/670067/186260954-5007d60f-19fd-47f9-95ed-89481ebec277.png">

sirbrillig reviewed (commented):
I don't have anything to add to the other reviews, but this approach seems to work well for checkout. 👍

manzoorwanijk committed:
Restore the previous behavior of atomic sites being treated as Jetpack sites

manzoorwanijk committed:
Revert "Revert the change made in #66828 which is not needed now"

This reverts commit 156c01aabe1815055e2f8feaed158f4dc68809d4.

manzoorwanijk commented:
> when I used an Atomic site to check for regressions to #66828, the button to upgrade Jetpack Search wasn't visible in the UI. I'm not yet sure why this is

Thank you @mattgawarecki for the review. I have fixed the issue. Actually, Atomic sites have been treated as Jetpack sites inside `isJetpackSite`. So, I have restored that behavior and added an extra option just in case someone doesn't want the Atomic sites to be treated as Jetpack sites using using `isJetpack` site, so that they don't need to use an extra check like `isAtomicSite()`.

I have reverted the changes that removed the changes made in #66828. Thus, that part should not be affected now. So, we are good to go.

manzoorwanijk committed:
Update docs

manzoorwanijk committed:
Update and enhance unit tests

mattgawarecki reviewed (approved):
@manzoorwanijk I tested again and everything appears to work correctly, including the banner in Settings -> Performance. I think this is good to ship, great work! 👍 🚢
